### PR TITLE
[WIP] Fetch trusted CA using SDS service.

### LINKED ISSUE #4763: SDS feature request:  only fetch trusted CA from SDS for CertificateValidationContext
*Title*: only fetch trusted CA from SDS for CertificateValidationContext

*Description*:
Istio likes to use SDS to send trusted CA for CertificateValidationContext.  But its SDS server doesn't have info for other fields, such as verify_subject_alt_name.  The discovery server has the info for "verify_subject_alt_name".

Like to change SDS that:
* Only trusted CA can be fetched from SDS,  the other fields of CertificateValidationContext can be specified in LDS/CDS from discovery server.
* Keep the current SDS of fetching whole CertificateValidationContext from SDS server.

Proposed Config change PR is:  https://github.com/envoyproxy/envoy/pull/4730

### PR DESCRIPTION
**Pull request by JimmyCYJ**
Description:
Support Envoy to fetch trusted CA from remote server via SDS API. This is the second part following PR #4730. 

Risk Level: Low
Testing: Unit tests and integration tests.
Docs Changes: NONE
Release Notes: NONE
Fixes #4763

Signed-off-by: JimmyCYJ jimmychen.0102@gmail.com

### CODE REVIEW

File: source/common/ssl/context_config_impl.h [hunk 1]
```diff
@@ -33,17 +33,28 @@ class ContextConfigImpl : public virtual Ssl::ContextConfig {
...
     // tls_certficate_provider_->secret() is NOT nullptr and
     // either certficate_validation_context_provider_ is nullptr or
-    // certficate_validation_context_provider_->secret() is NOT nullptr.
+    // certficate_validation_context_provider_->secret() is NOT nullptr and
+    // either trusted_ca_provider_ is nullptr or trusted_ca_provider_->secret() is NOT nullptr.
     return (!tls_certficate_provider_ || tls_certficate_provider_->secret() != nullptr) &&
            (!certficate_validation_context_provider_ ||
-            certficate_validation_context_provider_->secret() != nullptr);
+            certficate_validation_context_provider_->secret() != nullptr) &&
+           (!trusted_ca_provider_ || trusted_ca_provider_->secret() != nullptr);
```
Comment (qiwzhang) [inline source/common/ssl/context_config_impl.h]: This is not easy to read.  How about split it into two calls.   isTlsCertificateReady() and isValidationContextReady().  
IsRead() has to be both ready.

File: source/common/ssl/context_config_impl.h [hunk 2]
```diff
@@ -60,6 +71,12 @@ class ContextConfigImpl : public virtual Ssl::ContextConfig {
       cvc_update_callback_handle_ =
           certficate_validation_context_provider_->addUpdateCallback(callback);
     }
+    if (trusted_ca_provider_) {
+      if (tca_update_callback_handle_) {
+        tca_update_callback_handle_->remove();
+      }
+      tca_update_callback_handle_ = trusted_ca_provider_->addUpdateCallback(callback);
+    }
```
Comment (qiwzhang) [inline source/common/ssl/context_config_impl.h]: We have 3 handlers,  it is time to use a list for them.  Maybe define a struct to hold {provider_,  and handler_} and use a list to store them

File: api/envoy/api/v2/auth/cert.proto [hunk 3]
```diff
@@ -130,7 +130,13 @@ message CertificateValidationContext {
   //
   // See :ref:`the TLS overview <arch_overview_ssl_enabling_verification>` for a list of common
   // system CA locations.
-  core.DataSource trusted_ca = 1;
+  oneof trusted_ca_type {
```
[the remark this answers is not in the document]
Comment (lizan) [reply api/envoy/api/v2/auth/cert.proto]: I don't think I posted the doc, context:
https://docs.google.com/document/d/12gdjGN5m3v4vxUnDAglCP6pyyMoeuVGAGo7D_jc27jw/edit

[the remark this answers is not in the document]
Comment (JimmyCYJ) [reply api/envoy/api/v2/auth/cert.proto]: > In this PR, we enable the trusted CA to be split out, but maybe we would want other fields also at some point, e.g. the SPKI.

With option (4), we can let SDS server to decide what subset of fields to send. That provides flexibility compared to this approach.

**Comment by lizan** [comment]
@JimmyCYJ Can we close this in favor of #4932?

**Pull request closed without merging**